### DISCUSSION TITLE
select(exists(MyModel)) marked as Optional, but always returns bool (mypy)

### DISCUSSION THREAD

**Discussion by Mastermind-U** [sqlalchemy/sqlalchemy#11987] [Usage Questions] [answered]
### Ensure stubs packages are not installed

- [X] No sqlalchemy stub packages is installed (both `sqlalchemy-stubs` and `sqlalchemy2-stubs` are not compatible with v2)

### Verify if the api is typed

- [X] The api is not in a module listed in [#6810](https://github.com/sqlalchemy/sqlalchemy/issues/6810) so it should pass type checking

### Describe the typing issue

Both `select(exists(MyModel))` and select(select(MyModel).exists()) scalar result marked as `bool | None`, but it returns `bool` value in 100% cases.

### To Reproduce

```python
import asyncio

from sqlalchemy import exists, select
from sqlalchemy.ext.asyncio import (
    AsyncAttrs,
    AsyncSession,
    async_sessionmaker,
    create_async_engine,
)
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


class Base(DeclarativeBase, AsyncAttrs):
    pass


class MyModel(Base):
    __tablename__ = "mytable"
    id: Mapped[int] = mapped_column(primary_key=True)


async def does_rows_exists(session: AsyncSession) -> bool:
    return await session.scalar(select(exists(MyModel)))


async def main() -> None:
    engine = create_async_engine("sqlite+aiosqlite://", echo=True)
    async_session = async_sessionmaker(engine, expire_on_commit=False)

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    async with async_session() as session:
        result = await does_rows_exists(session)
        print(result)
        assert isinstance(result, bool)


asyncio.run(main())
```

### Error

```sh
$ mypy somefile.py
error: Incompatible return value type (got "bool | None", expected "bool")  [return-value]
```

### Versions

- OS: Debian (python:3.12.6-bookworm)
- Python: 3.12.6
- SQLAlchemy: {extras = ["asyncio", "mypy"], version = "2.0.35"}
- Type checker (eg: mypy 0.991, pyright 1.1.290, etc): mypy 1.10.1 (compiled: no)

### Additional context

Config:
```toml
[tool.mypy]
plugins = ["sqlalchemy.ext.mypy.plugin"]
ignore_missing_imports = true
platform = "linux"
show_column_numbers = true
pretty = true
warn_unreachable = true
color_output = true
warn_unused_ignores = true
```

To fix this, you can use `bool(await session.scalar(select(exists(MyModel))))` or `cast(bool, await session.scalar(select(exists(MyModel)))))`, but i think this is a crutch.

**Comment by CaselIT** [comment]
Hi,

That's not correct.
Example:
```py
assert session.scalar(sa.select(sa.exists(1)).where(sa.false())) is None
```

**Reply by CaselIT** [reply]
Also for this you can use `session.scalars(query).one()`

**Comment by zzzeek** [comment, marked as answer]
> Both select(exists(MyModel)) and select(select(MyModel).exists()) scalar result marked as bool | None, 

this is incorrect.  Both of these statements are typed correctly as `Select[Tuple[bool]]`:

```py
async def does_rows_exists(session: AsyncSession) -> bool:

    stmt = select(exists(MyModel))
    reveal_type(stmt)

```

mypy:

```
]$ mypy test3.py 
test3.py:25: note: Revealed type is "sqlalchemy.sql.selectable.Select[tuple[builtins.bool]]"
Success: no issues found in 1 source file

```

but it returns bool value in 100% cases.

What you are seeing is that `session.scalar()` is typed as `Optional[_T]`, which is also correct; it doesnt know anything about how many rows the statement marked by `_T` might return, which can always be zero rows and for which the method will return `None`.

What you're looking for is [one()](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Result.one) that is typed as `_T` and will raise if there are no rows.  This is on the `Result` object and there is not currently a convenience form on sessions, so the correct code is:

```py
async def does_rows_exists(session: AsyncSession) -> bool:

    stmt = select(exists(MyModel))
    return (await session.scalars(stmt)).one()

```
